# 01. Limpeza dos dados
---

Primeiramente, vamos importar os dados, lendo o arquivo JSON disponibilizado via API.

Os dados foram fornecidos em um arquivo no formato JSON. Podemos importar diretamente em um DataFrame Pandas, usando o método read_json(), ou podemos usar o método json.load() para importar como um dicionário.

Optei por começar pelo método json.load() para uma breve exploração em como os dados foram fornecidos pela API.

In [1]:
import json

# Importando a base de dados fornecida
dados = json.load(open('../Dados/Telco-Customer-Churn.json', 'r'))


In [2]:
# Verificando o tamanho da base de dados antes de qualquer manipulação
len(dados)

7267

In [3]:
def dict_types(x):
    ''' Recebe um dicionário e retorna um dicionário que relaciona a chave e o tipo. '''
    return {k: str(type(v)) if type(v) is not dict else dict_types(v) for k, v in x.items()}

In [4]:
# Imprimir o conjunto chave e tipo do primeiro item da base de dados
print(json.dumps(dict_types(dados[0]), indent=4))

{
    "customerID": "<class 'str'>",
    "Churn": "<class 'str'>",
    "customer": {
        "gender": "<class 'str'>",
        "SeniorCitizen": "<class 'int'>",
        "Partner": "<class 'str'>",
        "Dependents": "<class 'str'>",
        "tenure": "<class 'int'>"
    },
    "phone": {
        "PhoneService": "<class 'str'>",
        "MultipleLines": "<class 'str'>"
    },
    "internet": {
        "InternetService": "<class 'str'>",
        "OnlineSecurity": "<class 'str'>",
        "OnlineBackup": "<class 'str'>",
        "DeviceProtection": "<class 'str'>",
        "TechSupport": "<class 'str'>",
        "StreamingTV": "<class 'str'>",
        "StreamingMovies": "<class 'str'>"
    },
    "account": {
        "Contract": "<class 'str'>",
        "PaperlessBilling": "<class 'str'>",
        "PaymentMethod": "<class 'str'>",
        "Charges": {
            "Monthly": "<class 'float'>",
            "Total": "<class 'str'>"
        }
    }
}


> Observações:
>- Há 7267 elementos na base de dados;
>- A base de dados possui dicionários aninhados;
>- Todos os campos no dicionário de dados constam no primeiro elemento da base;

## Importar base de dados para um DataFrame Pandas
---

In [5]:
import pandas as pd

# Já importa a base de dados normalizando-a
dados = pd.json_normalize(data=dados, max_level=3, sep='_')

In [6]:
# Verificando a quantidade de elementos e colunas
dados.shape

(7267, 21)

In [7]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   customerID                 7267 non-null   object 
 1   Churn                      7267 non-null   object 
 2   customer_gender            7267 non-null   object 
 3   customer_SeniorCitizen     7267 non-null   int64  
 4   customer_Partner           7267 non-null   object 
 5   customer_Dependents        7267 non-null   object 
 6   customer_tenure            7267 non-null   int64  
 7   phone_PhoneService         7267 non-null   object 
 8   phone_MultipleLines        7267 non-null   object 
 9   internet_InternetService   7267 non-null   object 
 10  internet_OnlineSecurity    7267 non-null   object 
 11  internet_OnlineBackup      7267 non-null   object 
 12  internet_DeviceProtection  7267 non-null   object 
 13  internet_TechSupport       7267 non-null   objec

In [8]:
dados.isnull().sum()

customerID                   0
Churn                        0
customer_gender              0
customer_SeniorCitizen       0
customer_Partner             0
customer_Dependents          0
customer_tenure              0
phone_PhoneService           0
phone_MultipleLines          0
internet_InternetService     0
internet_OnlineSecurity      0
internet_OnlineBackup        0
internet_DeviceProtection    0
internet_TechSupport         0
internet_StreamingTV         0
internet_StreamingMovies     0
account_Contract             0
account_PaperlessBilling     0
account_PaymentMethod        0
account_Charges_Monthly      0
account_Charges_Total        0
dtype: int64

> Observações:
>- Contém os mesmos 7267 elementos;
>- Contém todas as colunas indicadas no dicionário de dados;
>- Num primeiro momento parece não haver nenhum valor nulo ou em branco;

In [9]:
pd.set_option('display.max_columns', 21)
dados.head()

customerID Churn customer_gender  customer_SeniorCitizen customer_Partner  \
0  0002-ORFBO    No          Female                       0              Yes   
1  0003-MKNFE    No            Male                       0               No   
2  0004-TLHLJ   Yes            Male                       0               No   
3  0011-IGKFF   Yes            Male                       1              Yes   
4  0013-EXCHZ   Yes          Female                       1              Yes   

  customer_Dependents  customer_tenure phone_PhoneService phone_MultipleLines  \
0                 Yes                9                Yes                  No   
1                  No                9                Yes                 Yes   
2                  No                4                Yes                  No   
3                  No               13                Yes                  No   
4                  No                3                Yes                  No   

  internet_InternetService internet_OnlineSecurity internet_OnlineBackup  \
0                      DSL                      No                   Yes   
1                      DSL                      No                    No   
2              Fiber optic                      No                    No   
3              Fiber optic                      No                   Yes   
4              Fiber optic                      No                    No   

  internet_DeviceProtection internet_TechSupport internet_StreamingTV  \
0                        No                  Yes                  Yes   
1                        No                   No                   No   
2                       Yes                   No                   No   
3                       Yes                   No                  Yes   
4                        No                  Yes                  Yes   

  internet_StreamingMovies account_Contract account_PaperlessBilling  \
0                       No         One year                      Yes   
1                      Yes   Month-to-month                       No   
2                       No   Month-to-month                      Yes   
3                      Yes   Month-to-month                      Yes   
4                       No   Month-to-month                      Yes   

  account_PaymentMethod  account_Charges_Monthly account_Charges_Total  
0          Mailed check                     65.6                 593.3  
1          Mailed check                     59.9                 542.4  
2      Electronic check                     73.9                280.85  
3      Electronic check                     98.0               1237.85  
4          Mailed check                     83.9                 267.4

In [10]:
dados['Churn'].value_counts()

No     5174
Yes    1869
        224
Name: Churn, dtype: int64

É possível observar que há registros com valores em branco na coluna 'Churn'.
Acredito que o mesmo seja verdade para outras colunas do tipo string.

In [11]:
dados.iloc[0]

customerID                     0002-ORFBO
Churn                                  No
customer_gender                    Female
customer_SeniorCitizen                  0
customer_Partner                      Yes
customer_Dependents                   Yes
customer_tenure                         9
phone_PhoneService                    Yes
phone_MultipleLines                    No
internet_InternetService              DSL
internet_OnlineSecurity                No
internet_OnlineBackup                 Yes
internet_DeviceProtection              No
internet_TechSupport                  Yes
internet_StreamingTV                  Yes
internet_StreamingMovies               No
account_Contract                 One year
account_PaperlessBilling              Yes
account_PaymentMethod        Mailed check
account_Charges_Monthly              65.6
account_Charges_Total               593.3
Name: 0, dtype: object

In [12]:
# dados.iloc[:,1:-2].select_dtypes(object) para selecionar apenas as colunas categóricas
# aplicando .value_counts() para cada coluna e armazenando num DataFrame para facilitar a visualização
pd.DataFrame({c: dados[c].value_counts() for c in dados.iloc[:,1:-2].select_dtypes(object)})

Churn  customer_gender  customer_Partner  \
                            224.0              NaN               NaN   
Bank transfer (automatic)     NaN              NaN               NaN   
Credit card (automatic)       NaN              NaN               NaN   
DSL                           NaN              NaN               NaN   
Electronic check              NaN              NaN               NaN   
Female                        NaN           3592.0               NaN   
Fiber optic                   NaN              NaN               NaN   
Mailed check                  NaN              NaN               NaN   
Male                          NaN           3675.0               NaN   
Month-to-month                NaN              NaN               NaN   
No                         5174.0              NaN            3749.0   
No internet service           NaN              NaN               NaN   
No phone service              NaN              NaN               NaN   
One year                      NaN              NaN               NaN   
Two year                      NaN              NaN               NaN   
Yes                        1869.0              NaN            3518.0   

                           customer_Dependents  phone_PhoneService  \
                                           NaN                 NaN   
Bank transfer (automatic)                  NaN                 NaN   
Credit card (automatic)                    NaN                 NaN   
DSL                                        NaN                 NaN   
Electronic check                           NaN                 NaN   
Female                                     NaN                 NaN   
Fiber optic                                NaN                 NaN   
Mailed check                               NaN                 NaN   
Male                                       NaN                 NaN   
Month-to-month                             NaN                 NaN   
No                                      5086.0               707.0   
No internet service                        NaN                 NaN   
No phone service                           NaN                 NaN   
One year                                   NaN                 NaN   
Two year                                   NaN                 NaN   
Yes                                     2181.0              6560.0   

                           phone_MultipleLines  internet_InternetService  \
                                           NaN                       NaN   
Bank transfer (automatic)                  NaN                       NaN   
Credit card (automatic)                    NaN                       NaN   
DSL                                        NaN                    2488.0   
Electronic check                           NaN                       NaN   
Female                                     NaN                       NaN   
Fiber optic                                NaN                    3198.0   
Mailed check                               NaN                       NaN   
Male                                       NaN                       NaN   
Month-to-month                             NaN                       NaN   
No                                      3495.0                    1581.0   
No internet service                        NaN                       NaN   
No phone service                         707.0                       NaN   
One year                                   NaN                       NaN   
Two year                                   NaN                       NaN   
Yes                                     3065.0                       NaN   

                           internet_OnlineSecurity  internet_OnlineBackup  \
                                               NaN                    NaN   
Bank transfer (automatic)                      NaN                    NaN   
Credit card (automatic)                        NaN                    NaN   
DSL        

In [13]:
# Calcular a quantidade de valores vazios nas colunas
dados.isin(['', ' ']).sum()

customerID                     0
Churn                        224
customer_gender                0
customer_SeniorCitizen         0
customer_Partner               0
customer_Dependents            0
customer_tenure                0
phone_PhoneService             0
phone_MultipleLines            0
internet_InternetService       0
internet_OnlineSecurity        0
internet_OnlineBackup          0
internet_DeviceProtection      0
internet_TechSupport           0
internet_StreamingTV           0
internet_StreamingMovies       0
account_Contract               0
account_PaperlessBilling       0
account_PaymentMethod          0
account_Charges_Monthly        0
account_Charges_Total         11
dtype: int64

> Observações:
> - As colunas **Churn** e **account_Charges_Total** possuem valores em branco;
> - A coluna **Churn** é nossa coluna alvo, importante para o desenvolvimento de um modelo de Machine Learning, portanto, valor vazio nessa coluna irá atrapalhar o modelo. Removeremos esses registros.

In [14]:
# Excluindo registros com **Churn** vazio
dados.drop(dados[dados['Churn']== ''].index, axis=0, inplace=True)

In [15]:
dados.shape

(7043, 21)

> Observações:
> - A coluna **account_Charges_Total** aparentemente é calculada à partir das colunas **account_Charges_Monthly** e **customer_tenure**.

In [16]:
# Convertendo a string ' ' para None
dados.loc[dados['account_Charges_Total'] == ' ', 'account_Charges_Total'] = None

In [17]:
# Convertendo a coluna account_Charges_Total para Float
dados['account_Charges_Total'] = dados['account_Charges_Total'].astype(float)

In [18]:
# Verificando a correção entre as colunas numéricas
dados.corr(numeric_only=True)

customer_SeniorCitizen  customer_tenure  \
customer_SeniorCitizen                 1.000000         0.016567   
customer_tenure                        0.016567         1.000000   
account_Charges_Monthly                0.220173         0.247900   
account_Charges_Total                  0.102411         0.825880   

                         account_Charges_Monthly  account_Charges_Total  
customer_SeniorCitizen                  0.220173               0.102411  
customer_tenure                         0.247900               0.825880  
account_Charges_Monthly                 1.000000               0.651065  
account_Charges_Total                   0.651065               1.000000

In [19]:
# Verificando os registros no qual account_Charges_Total é vazio
dados[dados['account_Charges_Total'].isna()]

customerID Churn customer_gender  customer_SeniorCitizen  \
975   1371-DWPAZ    No          Female                       0   
1775  2520-SGTTA    No          Female                       0   
1955  2775-SEFEE    No            Male                       0   
2075  2923-ARZLG    No            Male                       0   
2232  3115-CZMZD    No            Male                       0   
2308  3213-VVOLG    No            Male                       0   
2930  4075-WKNIU    No          Female                       0   
3134  4367-NUYAO    No            Male                       0   
3203  4472-LVYGI    No          Female                       0   
4169  5709-LVOEQ    No          Female                       0   
5599  7644-OMVMY    No            Male                       0   

     customer_Partner customer_Dependents  customer_tenure phone_PhoneService  \
975               Yes                 Yes                0                 No   
1775              Yes                 Yes                0                Yes   
1955               No                 Yes                0                Yes   
2075              Yes                 Yes                0                Yes   
2232               No                 Yes                0                Yes   
2308              Yes                 Yes                0                Yes   
2930              Yes                 Yes                0                Yes   
3134              Yes                 Yes                0                Yes   
3203              Yes                 Yes                0                 No   
4169              Yes                 Yes                0                Yes   
5599              Yes                 Yes                0                Yes   

     phone_MultipleLines internet_InternetService internet_OnlineSecurity  \
975     No phone service                      DSL                     Yes   
1775                  No                       No     No internet service   
1955                 Yes                      DSL                     Yes   
2075                  No                       No     No internet service   
2232                  No                       No     No internet service   
2308                 Yes                       No     No internet service   
2930                 Yes                      DSL                      No   
3134                 Yes                       No     No internet service   
3203    No phone service                      DSL                     Yes   
4169                  No                      DSL                     Yes   
5599                  No                       No     No internet service   

     internet_OnlineBackup internet_DeviceProtection internet_TechSupport  \
975                    Yes                       Yes                  Yes   
1775   No internet service       No internet service  No internet service   
1955                   Yes                        No                  Yes   
2075   No internet service       No internet service  No internet service   
2232   No internet service       No internet service  No internet service   
2308   No internet service       No internet service  No internet service   
2930                   Yes                       Yes                  Yes   
3134   No internet service       No internet service  No internet service   
3203                    No                       Yes                  Yes   
4169                   Yes                       Yes                   No   
5599   No internet service       No internet service  No internet service   

     internet_StreamingTV internet_StreamingMovies account_Contract  \
975                   Yes                       No         Two year   
1775  No internet service      No internet service         Two year   
1955                   No                       No         Two year   
2075  No internet service      No internet service         One year   
2232  No internet service      No i

In [20]:
# Verificando os registros no qual customer_tenure é 0
dados[dados['customer_tenure'] == 0]

customerID Churn customer_gender  customer_SeniorCitizen  \
975   1371-DWPAZ    No          Female                       0   
1775  2520-SGTTA    No          Female                       0   
1955  2775-SEFEE    No            Male                       0   
2075  2923-ARZLG    No            Male                       0   
2232  3115-CZMZD    No            Male                       0   
2308  3213-VVOLG    No            Male                       0   
2930  4075-WKNIU    No          Female                       0   
3134  4367-NUYAO    No            Male                       0   
3203  4472-LVYGI    No          Female                       0   
4169  5709-LVOEQ    No          Female                       0   
5599  7644-OMVMY    No            Male                       0   

     customer_Partner customer_Dependents  customer_tenure phone_PhoneService  \
975               Yes                 Yes                0                 No   
1775              Yes                 Yes                0                Yes   
1955               No                 Yes                0                Yes   
2075              Yes                 Yes                0                Yes   
2232               No                 Yes                0                Yes   
2308              Yes                 Yes                0                Yes   
2930              Yes                 Yes                0                Yes   
3134              Yes                 Yes                0                Yes   
3203              Yes                 Yes                0                 No   
4169              Yes                 Yes                0                Yes   
5599              Yes                 Yes                0                Yes   

     phone_MultipleLines internet_InternetService internet_OnlineSecurity  \
975     No phone service                      DSL                     Yes   
1775                  No                       No     No internet service   
1955                 Yes                      DSL                     Yes   
2075                  No                       No     No internet service   
2232                  No                       No     No internet service   
2308                 Yes                       No     No internet service   
2930                 Yes                      DSL                      No   
3134                 Yes                       No     No internet service   
3203    No phone service                      DSL                     Yes   
4169                  No                      DSL                     Yes   
5599                  No                       No     No internet service   

     internet_OnlineBackup internet_DeviceProtection internet_TechSupport  \
975                    Yes                       Yes                  Yes   
1775   No internet service       No internet service  No internet service   
1955                   Yes                        No                  Yes   
2075   No internet service       No internet service  No internet service   
2232   No internet service       No internet service  No internet service   
2308   No internet service       No internet service  No internet service   
2930                   Yes                       Yes                  Yes   
3134   No internet service       No internet service  No internet service   
3203                    No                       Yes                  Yes   
4169                   Yes                       Yes                   No   
5599   No internet service       No internet service  No internet service   

     internet_StreamingTV internet_StreamingMovies account_Contract  \
975                   Yes                       No         Two year   
1775  No internet service      No internet service         Two year   
1955                   No                       No         Two year   
2075  No internet service      No internet service         One year   
2232  No internet service      No i

In [21]:
dados.loc[dados['account_Charges_Total'].isna(), 'account_Charges_Total'] = dados['account_Charges_Monthly']

> Observações:
> - Haviam 11 registros sem valor na coluna **account_Charges_Total**;
> - Nos casos em que a coluna **account_Charges_Total** era 0, a coluna **customer_tenure** também era 0, e vice-versa.
> - Estes registros seriam o equivalente aos novos clientes, que ainda não realizaram nenhum pagamento.
> - Foi considerado o valor mensal do serviço como valor total nestes casos.

In [22]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7043 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   customerID                 7043 non-null   object 
 1   Churn                      7043 non-null   object 
 2   customer_gender            7043 non-null   object 
 3   customer_SeniorCitizen     7043 non-null   int64  
 4   customer_Partner           7043 non-null   object 
 5   customer_Dependents        7043 non-null   object 
 6   customer_tenure            7043 non-null   int64  
 7   phone_PhoneService         7043 non-null   object 
 8   phone_MultipleLines        7043 non-null   object 
 9   internet_InternetService   7043 non-null   object 
 10  internet_OnlineSecurity    7043 non-null   object 
 11  internet_OnlineBackup      7043 non-null   object 
 12  internet_DeviceProtection  7043 non-null   object 
 13  internet_TechSupport       7043 non-null   objec